src/gui/widgets/ControlLayout.cpp

```diff
@@ -0,0 +1,308 @@
+	  m_searchBar(new QLineEdit(parent))
+{
+	setContentsMargins(margin, margin, margin, margin);
+	m_searchBar->setPlaceholderText("filter");
+	m_searchBar->setObjectName(s_searchBarName);
+	connect(m_searchBar, SIGNAL(textChanged(const QString&)),
+		this, SLOT(onTextChanged(const QString& )));
+	addWidget(m_searchBar);
+	m_searchBar->setHidden(true); // nothing to filter yet
+}
+
+ControlLayout::~ControlLayout()
+{
+	QLayoutItem *item;
+	while ((item = takeAt(0))) { delete item; }
+}
+
+void ControlLayout::onTextChanged(const QString&)
+{
+	invalidate();
+	update();
+}
+
+void ControlLayout::addItem(QLayoutItem *item)
+{
+	QWidget* widget = item->widget();
+	const QString str = widget ? widget->objectName() : QString("unnamed");
+	m_itemMap.insert(str, item);
+	invalidate();
+}
+
+int ControlLayout::horizontalSpacing() const
+{
+	if (m_hSpace >= 0) { return m_hSpace; }
+	else
+	{
+		return smartSpacing(QStyle::PM_LayoutHorizontalSpacing);
+	}
+}
+
+int ControlLayout::verticalSpacing() const
+{
+	if (m_vSpace >= 0) { return m_vSpace; }
+	else
+	{
+		return smartSpacing(QStyle::PM_LayoutVerticalSpacing);
+	}
+}
+
+int ControlLayout::count() const
+{
+	return m_itemMap.size() - 1;
+}
+
+QMap<QString, QLayoutItem*>::const_iterator
+ControlLayout::pairAt(int index) const
+{
+	if (index < 0) { return m_itemMap.cend(); }
+
+	auto skip = [&](QLayoutItem* item) -> bool
+	{
+		return item->widget()->objectName() == s_searchBarName;
+	};
+
+	QMap<QString, QLayoutItem*>::const_iterator itr = m_itemMap.cbegin();
+	for (; itr != m_itemMap.cend() && (index > 0 || skip(itr.value())); ++itr)
+	{
+		if(!skip(itr.value())) { index--; }
+	}
+	return itr;
+}
+
+// linear time :-(
+QLayoutItem *ControlLayout::itemAt(int index) const
+{
+	auto itr = pairAt(index);
+	return (itr == m_itemMap.end()) ? nullptr : itr.value();
+}
+
+QLayoutItem *ControlLayout::itemByString(const QString &key) const
+{
+	auto itr = m_itemMap.find(key);
+	return (itr == m_itemMap.end()) ? nullptr : *itr;
+}
+
+// linear time :-(
+QLayoutItem *ControlLayout::takeAt(int index)
+{
+	auto itr = pairAt(index);
+	return (itr == m_itemMap.end()) ? nullptr : m_itemMap.take(itr.key());
+}
+
+Qt::Orientations ControlLayout::expandingDirections() const
+{
+	return Qt::Orientations();
+}
+
+bool ControlLayout::hasHeightForWidth() const
+{
+	return true;
+}
+
+int ControlLayout::heightForWidth(int width) const
+{
+	int height = doLayout(QRect(0, 0, width, 0), true);
+	return height;
+}
+
+void ControlLayout::setGeometry(const QRect &rect)
+{
+	QLayout::setGeometry(rect);
+	doLayout(rect, false);
+}
+
+QSize ControlLayout::sizeHint() const
+{
+	return minimumSize();
+}
+
+QSize ControlLayout::minimumSize() const
+{
+	// original formula from Qt's FlowLayout example:
+	// get maximum height and width for all children.
+	// as Qt will later call heightForWidth, only the width here really matters
+	QSize size;
+	for (const QLayoutItem *item : qAsConst(m_itemMap))
```

Comment:
It seems like [`qAsConst()`](https://doc.qt.io/qt-5/qtglobal.html#qAsConst) is not available in Qt 5.6 which is the our minimum requirement.

Reply:
FYI, the Qt documentation says it's equivalent to a `const_cast`.